### ISSUE TITLE
Compex WLE900VX

### ISSUE THREAD

**Issue by Level-9** [rsa9000/atheepmgr#6]
Hi!

I have a Compex VLE900VX as found at this link:  https://compex.com.sg/shop/wifi-module/802-11ac-wave-1/wle900vx-2/
In case the first link does not work I bought mine at: https://teklager.se/en/products/router-components/wle900vx-wireless-wifi-kit

When I run the  the atheepmgr I encounter the following errors:
`me@wrap:~/src/atheepmgr$ sudo ./atheepmgr -v -P 1:0`
`Found Device: 168c:003c (QCA9860/QCA9862/QCA9880/QCA9882/QCA9890/QCA9892)`
`Try to map f7200000-f73fffff I/O region to the process memory`
`Mapped IO region at: 0x7f36cb65c000`
`Autodetected EEPROM map type is 9880`
`Atheros QCA9880 MAC/BB Rev:2 (SREV: 0x043202ff)`
`Unable to select GPIO access ops due to unsupported chip`
`Chip does not support EEPROM`
`OTP access ops: use QCA988x ops`
`Found block at 0: comp=3 ref=1 length=852 major=4 minor=4`
`can't find reference eeprom struct 1`
`Bad EEPROM length 0x53f0/0xf053 (expect 0x0844)`
`EEPROM check failed`
`Freeing Mapped IO region at: 0x7f36cb65c000`

Do I use the program wrong or is my card unsupported?

Relevant parts of output of lspci:

`01:00.0 Network controller: Qualcomm Atheros QCA986x/988x 802.11ac Wireless Network Adapter`
`05:00.0 Network controller: Qualcomm Atheros QCA986x/988x 802.11ac Wireless Network Adapter`

As you can see there are two cards in the computer.

Relevant part of syslog:
`Aug 25 13:44:00 wrap kernel: [   19.184833] ath: EEPROM regdomain: 0x0`
`Aug 25 13:44:00 wrap kernel: [   19.184853] ath: EEPROM indicates default country code should be used`
`Aug 25 13:44:00 wrap kernel: [   19.184882] ath: doing EEPROM country->regdmn map search`
`Aug 25 13:44:00 wrap kernel: [   19.184887] ath: country maps to regdmn code: 0x3a`
`Aug 25 13:44:00 wrap kernel: [   19.184891] ath: Country alpha2 being used: US`
`Aug 25 13:44:00 wrap kernel: [   19.184895] ath: Regpair used: 0x3a`

The underlying problem is that my EEPROM indicates that I'm in the US but I live in Sweden and my access point is now using frequencies not allowed here. Earlier I think one could use the ìw reg set` -command to but since upgrading to jammy this does not work. I was hoping to solve this by reprogram the EEPROM to indicate the correct country. Another option could be to compile a custom database for regdom but the database must be signed and this is complication things. The crda-util is, as I understand it - obsolete on Ubuntu Jammy.

I appreciate any guidance, thank you for your time and efforts.
Regards, Lars

**Comment by rsa9000** [comment]
Hello!

The template with id=1 is not supported at the moment. I have not added it yet since I have never seen such cards in the wild. I can do that, but it will not help in this particular case as atheepmgr does not support OTP memory writing and regdomain overwriting. Theoretically it is possible, but too dangerous.

CRDA has been deprecated for some time due to races in the regdomain configuration procedure. Now the whole regulatory database is loaded into the kernel at once by the firmware loading framework. Despite the CRDA deprecation, you still need the wireless-regdb package, which contains the stock database itself.

First check the regdb package installation:
`$ dpkg -l | grep wireless-regdb`

Then you can check the database load status:
`$ dmesg | grep regulatory`
`Loading firmware: regulatory.db`
`Loading firmware: regulatory.db.p7s`

You can force the database to be loaded with the following command (e.g. if you just install the package and do not want to reboot):
`# iw reg reload`

Once regulatory.db is loaded into the kernel, the `iw reg set SE` command should work as expected.

**Comment by Level-9** [comment]
Thank you for trying but it does not seem to work.
It is obviously totally unrelated to your code so feel free to close this issue.
However, if you find this intresting or want more info from my card (or for me to do any testing) - feel free to keep the thread open!

Regards, Lars

`me@wrap:~$ dpkg -l | grep wireless-regdb`
`ii  wireless-regdb                        2022.06.06-0ubuntu1~22.04.1                   all          wireless regulatory database`
`me@wrap:~$`
`me@wrap:~$ sudo dmesg | grep -i regula`
`[   17.003088] cfg80211: Loading compiled-in X.509 certificates for regulatory database`
`me@wrap:~$ sudo iw reg set SE`
`me@wrap:~$ sudo iw reg get`
`global`
`country 98: DFS-UNSET`
`       (2400 - 2472 @ 40), (N/A, 20), (N/A)`
`       (5150 - 5250 @ 100), (N/A, 23), (N/A), NO-OUTDOOR, AUTO-BW`
`       (5250 - 5350 @ 100), (N/A, 20), (0 ms), NO-OUTDOOR, DFS, AUTO-BW`
`       (5470 - 5725 @ 160), (N/A, 24), (0 ms), DFS`
`       (5725 - 5730 @ 5), (N/A, 13), (0 ms), DFS`
`       (5730 - 5850 @ 80), (N/A, 13), (N/A)`
`       (5850 - 5875 @ 25), (N/A, 13), (N/A), NO-OUTDOOR, PASSIVE-SCAN`
`       (5945 - 6425 @ 160), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN`
`       (57240 - 66000 @ 2160), (N/A, 40), (N/A)`

`phy#1`
`country US: DFS-FCC`
`       (902 - 904 @ 2), (N/A, 30), (N/A)`
`       (904 - 920 @ 16), (N/A, 30), (N/A)`
`       (920 - 928 @ 8), (N/A, 30), (N/A)`
`       (2400 - 2472 @ 40), (N/A, 30), (N/A)`
`       (5150 - 5250 @ 80), (N/A, 23), (N/A), AUTO-BW`
`       (5250 - 5350 @ 80), (N/A, 24), (0 ms), DFS, AUTO-BW`
`       (5470 - 5730 @ 160), (N/A, 24), (0 ms), DFS`
`       (5730 - 5850 @ 80), (N/A, 30), (N/A), AUTO-BW`
`       (5850 - 5895 @ 40), (N/A, 27), (N/A), NO-OUTDOOR, AUTO-BW, PASSIVE-SCAN`
`       (5925 - 7125 @ 320), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN`
`       (57240 - 71000 @ 2160), (N/A, 40), (N/A)`

`phy#0`
`country US: DFS-FCC`
`       (902 - 904 @ 2), (N/A, 30), (N/A)`
`       (904 - 920 @ 16), (N/A, 30), (N/A)`
`       (920 - 928 @ 8), (N/A, 30), (N/A)`
`       (2400 - 2472 @ 40), (N/A, 30), (N/A)`
`       (5150 - 5250 @ 80), (N/A, 23), (N/A), AUTO-BW`
`       (5250 - 5350 @ 80), (N/A, 24), (0 ms), DFS, AUTO-BW`
`       (5470 - 5730 @ 160), (N/A, 24), (0 ms), DFS`
`       (5730 - 5850 @ 80), (N/A, 30), (N/A), AUTO-BW`
`       (5850 - 5895 @ 40), (N/A, 27), (N/A), NO-OUTDOOR, AUTO-BW, PASSIVE-SCAN`
`       (5925 - 7125 @ 320), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN`
`       (57240 - 71000 @ 2160), (N/A, 40), (N/A)`

`me@wrap:~$ sudo iw reg reload`
`me@wrap:~$ sudo iw reg get`
`global`
`country 98: DFS-UNSET`
`	(2400 - 2472 @ 40), (N/A, 20), (N/A)`
`	(5150 - 5250 @ 100), (N/A, 23), (N/A), NO-OUTDOOR, AUTO-BW`
`	(5250 - 5350 @ 100), (N/A, 20), (0 ms), NO-OUTDOOR, DFS, AUTO-BW`
`	(5470 - 5725 @ 160), (N/A, 24), (0 ms), DFS`
`	(5725 - 5730 @ 5), (N/A, 13), (0 ms), DFS`
`	(5730 - 5850 @ 80), (N/A, 13), (N/A)`
`	(5850 - 5875 @ 25), (N/A, 13), (N/A), NO-OUTDOOR, PASSIVE-SCAN`
`	(5945 - 6425 @ 160), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN`
`	(57240 - 66000 @ 2160), (N/A, 40), (N/A)`

`phy#1`
`country US: DFS-FCC`
`	(902 - 904 @ 2), (N/A, 30), (N/A)`
`	(904 - 920 @ 16), (N/A, 30), (N/A)`
`	(920 - 928 @ 8), (N/A, 30), (N/A)`
`	(2400 - 2472 @ 40), (N/A, 30), (N/A)`
`	(5150 - 5250 @ 80), (N/A, 23), (N/A), AUTO-BW`
`	(5250 - 5350 @ 80), (N/A, 24), (0 ms), DFS, AUTO-BW`
`	(5470 - 5730 @ 160), (N/A, 24), (0 ms), DFS`
`	(5730 - 5850 @ 80), (N/A, 30), (N/A), AUTO-BW`
`	(5850 - 5895 @ 40), (N/A, 27), (N/A), NO-OUTDOOR, AUTO-BW, PASSIVE-SCAN`
`	(5925 - 7125 @ 320), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN`
`	(57240 - 71000 @ 2160), (N/A, 40), (N/A)`

`phy#0`
`country US: DFS-FCC`
`	(902 - 904 @ 2), (N/A, 30), (N/A)`
`	(904 - 920 @ 16), (N/A, 30), (N/A)`
`	(920 - 928 @ 8), (N/A, 30), (N/A)`
`	(2400 - 2472 @ 40), (N/A, 30), (N/A)`
`	(5150 - 5250 @ 80), (N/A, 23), (N/A), AUTO-BW`
`	(5250 - 5350 @ 80), (N/A, 24), (0 ms), DFS, AUTO-BW`
`	(5470 - 5730 @ 160), (N/A, 24), (0 ms), DFS`
`	(5730 - 5850 @ 80), (N/A, 30), (N/A), AUTO-BW`
`	(5850 - 5895 @ 40), (N/A, 27), (N/A), NO-OUTDOOR, AUTO-BW, PASSIVE-SCAN`
`	(5925 - 7125 @ 320), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN`
`	(57240 - 71000 @ 2160), (N/A, 40), (N/A)`

**Comment by rsa9000** [comment]
Lars, you are right. The discussion goes beyond the utility usage scope. So I close the issue. If you still need assistance, then, please, post your case to the linux-wireless mailing list linux-wireless@vger.kernel.org (https://wireless.wiki.kernel.org/en/developers/mailinglists).

**Comment by rsa9000** [comment]
@Level-9 I just pushed a change that introduces the 'generic' template support. It is still impossible to rewrite the regdomain, but at least you can check other card configuration params :)

If you have time and your ath10k driver is built with the debugfs support, please check if the utility decompress the calibration data correctly:
`# ./atheepmgr -v -P 1:0 unpack data-custom.bin`
`# cat /sys/kernel/debug/ieee80211/phy0/ath10k/cal_data > data-driver.bin`
`# md5sum data-custom.bin data-driver.bin`

The files should have a same digest.

**Comment by Level-9** [comment]
@rsa9000 everything seems to work fine and I can confirm that the checksums are the same.

`root@wrap:/sys/kernel/debug/ieee80211/phy0/ath10k# md5sum cal_data`
`fb129ac42d519a13deebc5fc86412059  cal_data`
`root@wrap:/sys/kernel/debug/ieee80211/phy0/ath10k# md5sum /home/me/src/atheepmgr/data-custom.bin `
`fb129ac42d519a13deebc5fc86412059  /home/me/src/atheepmgr/data-custom.bin`
`root@wrap:/sys/kernel/debug/ieee80211/phy0/ath10k#`

Thank you for supporting my card. It is much appreciated. Let me know if I can assist you by doing more testing.

Regards, Lars

**Comment by rsa9000** [comment]
Thank you Lars! While I was importing the generic template, I recalled another reason not to do it earlier. It looks sane, but has an invalid checksum. So I was wondering is this template found on a 3rd party site is valid or broken. And needs some confirmation, e.g. by comparison with the output of the firmware decompression code.

Now I can see that the template is ok and people can use the utility to audit their card configurations based on this template for sure. Thank you for testing and for sharing the results!

**Comment by Level-9** [comment]
I have yet another request to make. Could you please consider adding support and instructions for permanently updating the regdomain on the card? If that is possible and somewhat secure, it would be great. Including all necessary disclaimers of course.

None of the other found solutions to my problems appeals to me, because they will complicate my setup and probably stop working after upgrades.

One of the other possible solutions seems to be to apply the [OpenWrt-patch](https://github.com/openwrt/openwrt/blob/master/package/kernel/mac80211/patches/ath/402-ath_regd_optional.patch)  but that is not recommended according to [the known bugs and limitations section in this site](https://wireless.wiki.kernel.org/en/users/drivers/ath10k).

The other proposed solution is to create a custom wireless-regdb but that comes with all the associated problems mentioned earlier.

It seems like [more people](https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1895333) have this problem as well. Another card is mentioned there though, but else very much alike.

**Comment by rsa9000** [comment]
Regdomain updating is possible, but I think making such functionality public and easy to use is unsafe in many ways. So I do not want to implement it. I am sorry.

As for your card, I rechecked the posted regulatory configuration and found it to be correct. If we compare your configuration:

```
country 98: DFS-UNSET
        (2400 - 2472 @ 40), (N/A, 20), (N/A)
        (5150 - 5250 @ 100), (N/A, 23), (N/A), NO-OUTDOOR, AUTO-BW
        (5250 - 5350 @ 100), (N/A, 20), (0 ms), NO-OUTDOOR, DFS, AUTO-BW
        (5470 - 5725 @ 160), (N/A, 24), (0 ms), DFS
        (5725 - 5730 @ 5), (N/A, 13), (0 ms), DFS
        (5730 - 5850 @ 80), (N/A, 13), (N/A)
        (5850 - 5875 @ 25), (N/A, 13), (N/A), NO-OUTDOOR, PASSIVE-SCAN
        (5945 - 6425 @ 160), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN
        (57240 - 66000 @ 2160), (N/A, 40), (N/A)
```

with the Swedish regulation rules from wireless-regdb (https://git.kernel.org/pub/scm/linux/kernel/git/sforshee/wireless-regdb.git/tree/db.txt?id=9dc9c89375c996c4d1d1a8c0a76f9dd04e244077#n1508):

````
country SE: DFS-ETSI
        (2400 - 2483.5 @ 40), (100 mW)
        (5150 - 5250 @ 80), (200 mW), NO-OUTDOOR, AUTO-BW, wmmrule=ETSI
        (5250 - 5350 @ 80), (100 mW), NO-OUTDOOR, DFS, AUTO-BW, wmmrule=ETSI
        (5470 - 5725 @ 160), (500 mW), DFS, wmmrule=ETSI
        # short range devices (ETSI EN 300 440-1)
        (5725 - 5875 @ 80), (25 mW)
        # 6 GHz band
        (5945 - 6425 @ 160), (23), NO-OUTDOOR, wmmrule=ETSI
        # 60 GHz band channels 1-4 (ETSI EN 302 567)
        (57000 - 66000 @ 2160), (40)
````

we can see that your configuration is a subset of the country rules. In other words, your setup should in full compliance with the local radio regulation rules. Is that what you were looking for: restrict the operation of the AP to locally allowed channels? 

Or did I miss or misunderstand something? If so, then I propose moving this discussion to the linux-wireless mailing list. I am not an regulatory code expert. And I assume that the developers on the list can suggest a better solution or even make some changes to the driver/subsystem code to better support such WNICs. They just need a chance to know about such cases ;)

BTW, the '98' country in your report is a special id to indicate the intersection of the user-specified and earlier learned regulation rules. It could appear if the wireless subsystem learned the country code from a 3rd party beacon received during an earlier scan procedure. If you configure the country code immediately after booting and before starting the wireless interface, then the wireless subsystem will use the exact country that you manually specified.

**Comment by L2501** [comment]
ath10k allows loading the board data from file should be fine to dump the eeprom change the region and load it from file instead.
